cmd/gateway-gcs.go

```diff
@@ -375,19 +401,31 @@ func (l *gcsGateway) ListObjects(bucket string, prefix string, marker string, de
 		attrs, err := it.Next()
 		if err == iterator.Done {
 			break
-		} else if err != nil {
+		}
+		if err != nil {
 			return ListObjectsInfo{}, gcsToObjectError(traceError(err), bucket, prefix)
 		}
 
 		nextMarker = toGCSPageToken(attrs.Name)
 
-		if isGCSPrefix(attrs.Prefix) {
-			// we don't return our metadata prefix
+		if attrs.Prefix == gcsMinioPath {
+			// We don't return our metadata prefix.
 			continue
-		} else if attrs.Prefix != "" {
+		}
+		if !strings.HasPrefix(prefix, gcsMinioPath) {
+			// If client lists outside gcsMinioPath then we filter out gcsMinioPath/* entries.
+			// But if the client lists inside gcsMinioPath then we return the entries in gcsMinioPath/
+			// which will be helpful to observe the "directory structure" for debugging purposes.
+			if strings.HasPrefix(attrs.Prefix, gcsMinioPath) ||
+				strings.HasPrefix(attrs.Name, gcsMinioPath) {
```

Comment:
for what object name would `attrs.Name` have `gcsMinioPath`?
IIUC in `bucketName/minio.sys.temp/multipart/3318c0a5-551b-470a-8053-cb59a5668766/21b261593f1cf12f438d1dab975a421a`, `attrs.Prefix`  would be `minio.sys.temp/multipart/3318c0a5-551b-470a-8053-cb59a5668766` and `attrs.Name` would be `21b261593f1cf12f438d1dab975a421a`, right?